src/bms/parse/validity.rs

```diff
@@ -0,0 +1,465 @@
+        use std::collections::HashMap as StdHashMap;
+        let mut lane_to_notes: StdHashMap<(PlayerSide, Key), Vec<&Obj>> = StdHashMap::new();
+        for notes in self.notes.objs.values() {
+            for obj in notes {
+                // Visible note in section 000 (track index 0)
+                if obj.kind == NoteKind::Visible && obj.offset.track.0 == 0 {
+                    invalid.push(ValidityInvalid::VisibleNoteInTrackZero {
+                        side: obj.side,
+                        key: obj.key,
+                        time: obj.offset,
+                    });
+                }
+                lane_to_notes
+                    .entry((obj.side, obj.key))
+                    .or_default()
+                    .push(obj);
+            }
+        }
+        for ((side, key), objs) in lane_to_notes.into_iter() {
+            if objs.is_empty() {
+                continue;
+            }
+            // Sort by time
+            let mut lane_objs = objs;
+            lane_objs.sort_unstable_by_key(|o| o.offset);
+
+            // Build LN intervals by pairing consecutive Long notes
+            let mut long_times: Vec<ObjTime> = lane_objs
+                .iter()
+                .filter(|o| o.kind == NoteKind::Long)
+                .map(|o| o.offset)
+                .collect();
+            long_times.sort_unstable();
+            let mut ln_intervals: Vec<(ObjTime, ObjTime)> = Vec::new();
+            let mut iter = long_times.into_iter();
+            while let Some(start) = iter.next() {
+                if let Some(end) = iter.next() {
+                    if end >= start {
+                        ln_intervals.push((start, end));
+                    }
+                } else {
+                    break;
+                }
+            }
+
+            // Overlap single vs single at the same time
+            let mut i = 0;
+            while i < lane_objs.len() {
+                let time = lane_objs[i].offset;
+                let mut j = i;
+                let mut visible_count = 0usize;
+                while j < lane_objs.len() && lane_objs[j].offset == time {
+                    if lane_objs[j].kind == NoteKind::Visible {
+                        visible_count += 1;
+                    }
+                    j += 1;
+                }
+                if visible_count >= 2 {
+                    invalid.push(ValidityInvalid::OverlapVisibleSingleWithSingle {
+                        side,
+                        key,
+                        time,
+                    });
+                }
+                i = j;
+            }
+
+            // Helper: check if a time is within [s, e]
+            let time_overlaps_any_ln = |t: ObjTime| -> Option<(ObjTime, ObjTime)> {
+                for (s, e) in &ln_intervals {
+                    if t >= *s && t <= *e {
+                        return Some((*s, *e));
+                    }
+                }
+                None
```

Comment:
You can improve the performance like this algorithm with `long_times`:
1. Search the first item index `end_i` which is greater than or equal to `t` in `long_times` (or return `None` if not found).
    - Since `long_times` is sorted, it can be searched with [`slice::partition_point`](https://doc.rust-lang.org/std/primitive.slice.html#method.partition_point) rapidly like `long_times.partition_point(|&end| end < t)`.
2. Check whether `end_i` is an odd number. If it is not, there is no sections contains `t` so return `None`.
3. Return the section `(long_times[end_i - 1], long_times[end_i])`.